### ISSUE TITLE
Discussion: Support OTLP sending in addition to Remote Write

### ISSUE THREAD

**Issue by gouthamve** [prometheus/prometheus#12633]
### Proposal

As OTel gains more adoption, people might be interested in sending data from Prometheus to remote services through OTLP protocol in addition the Prometheus remote write protocol.

I think its a valid ask, and something that we should potentially consider. Thoughts?

**Comment by jaronoff97** [comment]
+1 to this! I would love to assist in making this possible if this is acceptable.

**Comment by scanfield** [comment]
+1 this would be great, would remove a lot of user pain and reduce complexity processing the WAL

**Comment by gouthamve** [comment]
> would remove a lot of user pain and reduce complexity processing the WAL

Which WAL are you talking about? Could you elaborate?

**Comment by scanfield** [comment]
Ah, I spoke too soon, I was thinking of an old version of this feature that tried to consume the prom WAL directly. But I do think that supporting OTLP directly would make things more flexible for users and folks building compatible systems.

**Comment by arukiidou** [comment]
Is this supported in agent mode?

That is, an agent that receives OTLP or prometheus and writes OTLP remotely

**Comment by bwplotka** [comment]
Nice, thanks for raising this!

I'm supportive, we discussed it a few times already. To be specific--[this means OTLP metric protocol part](https://github.com/open-telemetry/opentelemetry-proto/tree/main/opentelemetry/proto/metrics/v1)(: 

Just note that this might require current WAL tail + remote write sharding re-implementation to be redesigned for metric atomicity, as histograms are atomic on the OTLP (https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L384). Alternative this might only work with native histograms. 🙃 For initial implementation, no sharding implementation might work too.

Would be curious how this can evolve 💪🏽

**Comment by dodegaard** [comment]
+1 from someone explaining and implementing with customers this would be a huge positive for enablement.

**Comment by jaronoff97** [comment]
@gouthamve did someone do this already? if not, id be happy to give it a try!

**Comment by gouthamve** [comment]
Hi, this will be discussed tomorrow at the [Prometheus dev summit](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit) as part of "Prometheus becoming OTel native".

We don't yet have consensus that we want this in Prometheus, but I hope to get it tomorrow. I'll update the issue after the discussion tomorrow.

**Comment by jaronoff97** [comment]
@gouthamve I have some cycles this week to work on this. I saw the consensus here:
> CONSENSUS: We want to invite the OTel community to contribute patches enabling Prometheus to send OTLP metrics. Ideally some of them would join us as maintainers for the long term

Does that mean I'm okay to work on an alpha?

**Comment by cstyan** [comment]
I'm not fully understanding Bartek's comment regarding histogram atomicity, if someone from OTel could elaborate that would be helpful. Is it related to needing all series for a histogram (ie all buckets and their values) to be received at the same time? 

In general I'd be against a completely separate code path just to support OTLP, and suggest we go with either of his suggested workarounds of either a) only sending native histograms or b) finding a nice hack to disable sharding if we're sending OTLP. Transactional remote write (ie per scrape atomicity of sample sending) is still something we want to implement.

**Comment by npazosmendez** [comment]
Hey @jaronoff97, you may be interested in the work from https://github.com/prometheus/prometheus/pull/11640 . It is the intent to include per-sample metadata in [Remote Write 1.1](https://docs.google.com/document/d/1PljkX3YLLT-4f7MqrLt7XCVPG3IsjRREzYrUzBxCPV0/edit#heading=h.m5w2zqslg10y), which I understand should be useful for the work you are doing

**Comment by bwplotka** [comment]
👋🏽  

With the [Remote Write 2.0](https://prometheus.io/docs/specs/remote_write_spec_2_0/) spec released and some pieces for it merged (e.g. better metadata support in WAL), it might be time to re-evaluate adding [OTLP metrics](https://opentelemetry.io/docs/specs/otlp/) to Prometheus. I might have time for some quick PoC, but feel free to join the effort, we can pair!

The interesting parts to double check:

* *Sharding*: If we do no sharding, how practical this will be? Nevertheless, for PoC we can skip sharding.
* *Histogram atomicity*: If we don't do sharding, in theory we can skip this part. However, with PRW 2.0 we invented [custom native histograms (nhcb)](https://github.com/prometheus/proposals/blob/main/proposals/2024-01-26_classic-histograms-stored-as-native-histograms.md) that can help us here once we implement auto-conversion (something [todo for efficiency](https://github.com/prometheus/prometheus/issues/13532) anyway, see [discussion](https://groups.google.com/g/prometheus-developers/c/18xhBVm0_MA)).
* *Data Model*: We can follow [official guide](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#prometheus-metric-points-to-otlp), plus we could borrow [`prometheusreceiver` code that does similar](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0f2b59ac243be499c34f57e24f5e7203c96f1c45/receiver/prometheusreceiver/internal/transaction.go#L85).
* *Naming:* The current [translation guidance](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/prometheus/README.md#list-of-transformations-performed-to-convert-prometheus-metrics-to-opentelemetry-metrics) looks simple, but obviously we will break semantic convention.
* *Start Time/Created Timestamp:* This might be the hardest piece as start time looks like [a required feature](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#start-time) as opposed to [Remote Write 2.0 optional CT](https://prometheus.io/docs/specs/remote_write_spec_2_0/#:~:text=created_timestamp%20SHOULD%20be%20provided%20for%20metrics%20that%20follow%20counter%20semantics%20(e.g.%20counters%20and%20histograms).%20Receivers%20MAY%20reject%20those%20series%20without%20created_timestamp%20being%20set.). The general plan for this work is already there and tracked [here](https://github.com/prometheus/prometheus/issues/14217).

What am I missing? 🤗

**Comment by jaronoff97** [comment]
@bwplotka i would love to pair on this and get a PoC out there! Maybe we can meet next week?

**Comment by ywwg** [comment]
adding more thoughts -- it looks like there are two main categories of normalization for names -- technical, and semantic. There are technical changes like changing dots to underscores and other cleanups of the names so that they work with Prometheus, and there are the semantic changes to add suffixes for units and types.  Would it make sense to make those two separate options so users can pick if they want one or both sets of normalization, rather than always insisting on the technical normalization with optional semantic normalization?

**Comment by dashpole** [comment]
> obviously we will break semantic convention

Note that going from Prometheus -> OTel should be easier than the other way, since OTel is generally more permissive.  We shouldn't need to do any renaming on this path.

**Comment by ArthurSens** [comment]
Hey folks, I bring news from our recent discussion that happened during our in-person Prometheus DevSummit in Berlin.

**TLDR; Prometheus will not support exporting OTLP data, at least not for now.**

Long answer:


Prometheus Remote-Write v2(PRWv2) is just out of the press! @bwplotka and @alexgreenbank have [presented the results of the work during PromCon](https://www.youtube.com/watch?v=drSSR8RE2jM&t=4542s). The benchmark results are impressive:

<img width="1319" alt="image" src="https://github.com/user-attachments/assets/1386ee78-5fe7-4301-98b0-d5b4dde7ab9b">

As they mentioned in the talk, PRWv2 is marked as experimental and to call it stable we'd like to see adoption from 3-4 vendors. We believe that it will take too long if we wait for organic adoption, so the team is actively helping the ecosystem in this effort with Prometheus team members getting hands-on with other projects like Thanos/OTel-Collector/Mimir to implement the new version of the Protocol. Implementing and supporting yet another push protocol could harm the adoption of PRWv2 and that's something we'd like to avoid since it's the most suitable protocol for Prometheus metrics.

However, we don't want to ignore that there are people who have instrumented their applications with Prometheus SDKs, use Prometheus as their tool for metric collection, and would like to send those metrics somewhere else, e.g. cloud offerings,  that doesn't support receiving metrics via Remote Write. Those are the users who need a way to export OTLP data from Prometheus the most! We see this as a symptom that the PRWv1 format was incompatible with the OTLP format, mostly around Histograms where the former could potentially send buckets of the same histogram in different requests, while the latter requires histograms to be atomic. PRWv2 solves this problem and therefore implementing a `prometheusremotewritereceiver` in the OpenTelemetry-Collector is now possible. The [Prometheus-Otel Working Group](https://cloud-native.slack.com/archives/C01LSCJBXDZ), with members from both communities (including me), will implement and maintain this component, closing the gap:

```mermaid
---
title: Prometheus metrics to OTLP vendor
---

flowchart TB
  promsdk("App instrumented with Prometheus SDK")
  prom[("Prometheus")]
  col("Otel-Collector")
  vendor[("Vendor")]
  promsdk -- Scrape --> prom;
  prom -- Remote-Write v2 --> col;
  col -- OTLP --> vendor;
```

With that said, I'm closing the issue for now.

We're open to the idea of restarting this conversation if new-use cases appear that can't be solved with the OpenTelemetry-Collector!

**Comment by bwplotka** [comment]
For context on how this decision was made, see [the relevant DevSummit discussion](https://docs.google.com/document/d/1uurQCi5iVufhYHGlBZ8mJMK_freDFKPG0iYBQqJ9fvA/edit#bookmark=kix.v1ebh21vfcjc).

**Comment by bwplotka** [comment]
Additionally, feedback welcome, if you are end-user, and this decision impacted you in any way!

**Comment by jaronoff97** [comment]
For now, I think this should be okay by me! The reasoning here makes sense from an (otel or prometheus) developer perspective as getting to the initial goal of easier connection between these components should be more straightforward with this new protocol. After PRWv2 is fully supported and adopted by more organizations, if we still need/want OTLP writing we can evaluate this later. For now, I'll be happy to work with @ArthurSens and @bwplotka on pushing the PRWv2 side forward in the otel world. Thank you both for your great work here :bow:
